### DISCUSSION TITLE
Possibility to do a font toggle button?

### DISCUSSION THREAD

**Discussion by Lazorne** [squidfunk/mkdocs-material#6512] [Q&A]
An idea that I had was I wanted a button on our MKDocs Material site that changes the font for that that users session with the press of a button.

The use-case is accessibility:

I want users with dyslexia to have an easy toggle to switch the entire site over to the free and open source OpenDyslexic font from:

https://opendyslexic.org/

Is this possible to do?

Thanks in advance :)

**Comment by alexvoss** [comment]
It should be possible to do as a customization and I would love to help as I think this is an important thing to do and an excellent way to highlight how well-designed and flexible Material for MkDocs is. It might take a bit of time and I am not entirely sure when I could start on this. In any case, this would be great to work on collaboratively, so, folks, LMK if you are interested.

**Reply by kamilkrzyskow** [reply]
I've been slowly working on a preferences page for the GMC website, which could allow to change the base color tint of the website via JavaScript (e.g. blue -> orange). 
I have hit a kind of wall in finding the best way to convert HEX to HSL directives.

Getting back on topic, I could also include a font changing mechanism 🤔 as it is also managed via stylesheets.

However, I don't think I will hook into the Google Fonts API to allow the user to search for their favourite font. I would probably setup a selection input with predefined fonts.

Edit: Of course changing to an OpenDyslexic font is also a possibility.

**Reply by alexvoss** [reply]
If we find enough common ground we could pull this off together, I am sure. I would love to see more customizations being traded in the community in general. Especially since the introduction of hooks in MkDocs there is so much scope for stuff that does not have to be a fully-fledged plugin but can add real value to the ecosystem. **Edit**: do you have your code for the preference page somewhere and would you be willing to or have you already licensed it under something like MIT?

**Reply by kamilkrzyskow** [reply]
I don't mind sharing the code @alexvoss or making it under the MIT licence, but it's not a hook this time. 😅 
I've pushed the branch here https://github.com/kamilkrzyskow/gmc/tree/preferences-page only the [last commit](https://github.com/kamilkrzyskow/gmc/commit/d57a0ae87b2836a008513548ce262c256903ee36) is relevant.
There is also a preview here: https://kamilkrzyskow.github.io/gmc/preferences/

The `preferences.md` file includes a JSON structure for localization, and the script tag to the `preferences.js` file with an absolute src uri. 
The `preferences.js` file creates the toggle elements and attaches logic to them before the `bundle.js` hooks the tooltips titles. I guess this isn't really pro-accessibility 🤔 for screen readers etc., so perhaps the creation of the elements should be a separate `preferences.html` template file.
The `main.html` file overrides the `extrahead` block adding some global JavaScript variables, and contains the logic to load the CSS styles from localStorage.

It's heavily work in progress, so it's rather bare-bones at the moment. I'm also not sure how the collaboration would look here TBH.
When I started working on it, the idea was to be able to quickly override styles, but now as the accessibility idea came about I have to rethink this. As you said for the fonts a hook would've been better as it would allow to download font files during build, or should we leave such things for the `privacy` plugin?

EDIT:
When finished, this could also go to the examples repository, as a more elaborate example of customizing the docs with JavaScript/Templates/Block Overrides.

**Reply by ghost** [reply]
I took @kamilkrzyskow's code and separated out only the font part, putting it in the page header.

![image](https://github.com/user-attachments/assets/ae7fdb64-fbd2-44f9-9baf-be61605d9997)

These options are configured in **mkdocs.yml**:

``` yaml
fonts:
  icon: material/format-font
  preferences:
    - text: Default
    - text: Roboto
      code: Roboto Mono
    - text: OpenDyslexic
      code: OpenDyslexicMono
      src: https://fonts.cdnfonts.com/css/opendyslexic
```

If you want to take a look or even implement it, take a look at the PR [dunossauro/fastapi-do-zero#211](https://github.com/dunossauro/fastapi-do-zero/pull/221):
You can see the results at https://fastapidozero.dunossauro.com/.

**Reply by kamilkrzyskow** [reply]
Nice work @taconi, perhaps add this discussion in a comment in your JavaScript source? For people to have more context.
When switching to the OpenDyslexic font there is a period of time when the page is blank, displays no text, this happens in my version too.

I think this happens because the defined fonts in the CSS file do not contain `font-display: swap;`:
```css
@font-face {
  font-family:opendyslexic;
  font-style:normal;
  font-weight:400;
  src:local('OpenDyslexic'),
  url(https://fonts.cdnfonts.com/s/19808/OpenDyslexic-Regular.woff) format('woff')
}
```
https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display

I happened to stumble upon this issue when working on the [Nype](https://npe.cm/) website:
- https://github.com/nypesap/nypesap.github.io/commit/1994d9a72b54607907549ade8855b8a2eea30d19

And this setting did help to mitigate this behaviour.

Perhaps you'd have to define your own CSS file with added `font-display` to fix the blank page issue.

**Reply by ghost** [reply]
@kamilkrzyskow sorry for the delay in responding.

> perhaps add this discussion in a comment in your JavaScript source?

The following function adds a trigger so that whenever a font is clicked a function can be called and switches to the selected font. 

``` javascript
(() => {
    const fonts = document.querySelectorAll("#md-fonts");
    fonts.forEach((font) => {
	let event_type = (font.tagName.toLowerCase() === "a") ? "click" : "change";
	font.addEventListener(event_type, updateFontPreference);
    });
})();
```

The `updateFontPreference` function will make this font change by changing the css.


> When switching to the OpenDyslexic font there is a period of time when the page is blank, displays no text

Yes, indeed, the first time this happens, but after the font is downloaded it no longer happens, but it still seems like strange behavior.

> Perhaps you'd have to define your own CSS file with added font-display to fix the blank page issue.

The idea of ​​using javascript was to change the CSS dynamically, I don't know if this is possible with a CSS file.

> [...]  the defined fonts in the CSS file do not contain font-display: swap;

Interesting, I don't know about this attribute.

I added it to the `updateFontPreference` function and it solves the problem of the screen being without text for a moment.

<details>

<summary>function `updateFontPreference`</summary>

```javascript
function updateFontPreference(event) {
    console.debug(event);
    let fontFace = "";
    let cssRaw = "";
    let root = "";

    const text = event.target.dataset["text"];
    const code = event.target.dataset["code"];
    const src = event.target.dataset["src"];

    if (src !== "") {
        cssRaw += `@import url('${src}');`;
        fontFace += `src: url('${src}');`;
    }
    if (text !== "") {
        root += `--md-text-font: "${text}";`;
        fontFace += `font-family: "${text}";`;
    }
    if (code !== "") {
        root += `--md-code-font: "${code}";`;
    }

    if (fontFace !== "") {
        cssRaw += `@font-face{ ${fontFace} font-display: swap; }`
    }
    if (root !== "") {
        cssRaw += `:root { ${root} }`
    }

    // update style
    const loadedPreferences = __md_get(preferencesKey);
    loadedPreferences["__global"]["cssRaw"] = cssRaw;
    __md_set(preferencesKey, loadedPreferences);
    loadSetPreferences();
}
```

</details>

I applied the adjustment on this page: [https://foxdotchord.readthedocs.io/](https://foxdotchord.readthedocs.io/)
The complete code can be seen here: https://codeberg.org/taconi/FoxDotChord/src/branch/main/overrides

> For people to have more context.

The idea was to create the possibility for people to define which fonts can be selected by the user through the `mkdocs.yaml` file:

``` yaml
fonts:
  icon: material/format-font
  preferences:
    - text: Roboto
      code: Roboto Mono
    - text: OpenDyslexic
      code: OpenDyslexicMono
      src: https://fonts.cdnfonts.com/css/opendyslexic
```

And the page reader would define which font they prefer by the header icon:
![image](https://github.com/user-attachments/assets/eca591ca-1000-4f9e-b8dc-b08a1f5d873a)


--- 

I don't know this is the best solution but the interface for the reader and for defining the sources in `mkdocs.yaml` pleases me.

Any point that can be improved or any other possible implementation will be of great help (especially for me who doesn't know much about front-end development).

**Comment by Lazorne** [comment]
I'm happy my little question got this much traction!

For the record I just started using MKDocs Material when I tried to design our wiki so I'm just very very new to this :) 
https://retrodeck.readthedocs.io/en/latest/ 

Right now you have native Google Font support, but any other font can be loaded via the additional fonts function. 

But I kinda feel that opendyslexic should get the same treatment as the Google Fonts without the need to go into additional fonts and specify custom sources. 
 
Is it possible to include [opendyslexic github link](https://github.com/antijingoist/opendyslexic) as a part of MKDocs Material so you can just do:

```
theme:
  font:
    code: opendyslexic
```

That would make it easier for some that just want to do a MKDocs Material site for just those people with dyslexia.

If shipped with MKDocs Material there also could be a quick function to just enable a open opendyslexic button for any site as one quick function and another function that you can have a button to toggle between any font you want from google or a different source. 

Also thanks in advance everyone! ❤️

**Reply by squidfunk** [reply]
IMHO, an even better idea that would benefit even more people would be submitting OpenDyslexic to Google Fonts. We're currently not considering integrating specific fonts, because if we start doing this, further requests from other users will follow. Please understand that we are a very small team that manages this project, so we need to balance what we can commit to maintain.

**Reply by Lazorne** [reply]
Understood and I can see why you don't want to open that can-of-worms with font submissions.
I will do start an issue over at the OpenDyslexic github with a suggestion to submit the font to Google Fonts.  

Let's see what happens!

Thanks for the replies :)

**Reply by squidfunk** [reply]
Awesome! You're invited to keep us updated ☺️

**Reply by Lazorne** [reply]
https://github.com/antijingoist/opendyslexic/issues/85

Let's see how it goes! 

🚂 🚃 🚃

The issue train keeps going. Let's hope we can make it the web a more accessible place from this!

**Reply by kamilkrzyskow** [reply]
Continuing after https://github.com/squidfunk/mkdocs-material/discussions/6512#discussioncomment-7867461

Here is a temporary deployment of the new preference feature in our docs over at Gothic Modding Community: 
https://kamilkrzyskow.github.io/gmc/preferences/
Commit: https://github.com/kamilkrzyskow/gmc/commit/e9e3e15d216237fce6564e041fd6ab4ee04e5257

There is a font change toggle with the OpenDyslexic font from https://www.cdnfonts.com/opendyslexic.font
This CDN seems to use some older version of https://github.com/antijingoist/opendyslexic or some fork of it, as it contains a Mono variant which isn't included in the official branch. Maybe you have some comments about that @Lazorne?

I'm mostly finished working on the preference implementation, mostly because I'm conflicted about how to optimize it further, so I should be able to extract some of it and prepare an example in https://github.com/mkdocs-material/examples for people to copy-paste into their own docs.

Known issues related to the font, which I'm still thinking about, but will likely leave it as they are now:
- The selection toggle injects an `@import` CSS statement to load after the load of the page. The import statement then downloads the CSS and sets the font. Therefore there is a delay when loading the custom font. To circumvent that, my only suggestion is to inject the contents of the `@imported` file directly, to avoid having to download it each time.
- The `<selection>` or `<textarea>` aren't styled as the rest of the theme. So the changed font / preferences don't affect them.

I opted out of creating the elements with JavaScript, as there was missing TOC or other stuff wasn't hooked with the built-in bundle.js.
I tried the template preferences.html approach, but it had similar issues.
So I opted into a mix of Markdown with HTML + JavaScript, as this seems to be the simplest. The i18n approach with the preferences.py hook is completely optional. 
The hook could be further extended to generate the HTML as well, or downloading the CDN `@import` file to replace the `@import` statement with its contents, but I feel that this is on the edge of reinventing the wheel 🤔 

Any feedback is welcome 😄 @alexvoss as the maintainer of the examples repository, do you have any suggestions of how to best minify it for others? 🤔

**Reply by alexvoss** [reply]
Hi @kamilkrzyskow, this is pretty cool. I like the idea of a user preference pane. It could also store a preference for the language and other things. Couple of other things:

- not sure why the override of `main.html` is needed? 
- the handling of the font turned out to be tricky. I would simplify and load the font by default to simplify the example
- perhaps I would remove the custom CSS bit. Anyone who caters for users that sophisticated probably would not struggle to add this back in?
- perhaps there are other design decisions that would not be necessary for the example?

I like how this shows a number of strong points of Material and MkDocs. IMHO would be a valuable addition to the examples. Work on that has stalled a little since starting to work on it raised a number of questions we are busy answering first. With the tag plugin revamped I am keen to have a go at it again, though. Now just to find the time...

**Reply by kamilkrzyskow** [reply]
Thanks for the feedback @alexvoss! As for your comments:
> not sure why the override of main.html is needed?

I believe it's needed to inject JavaScript code into the `<head>`, so that it loads the preference before the DOM was completely drawn. This contributes to avoiding any unnecessary UI changes / flashes / pop-in after the CSS was loaded, because the injected CSS is already taken into account. Similar approach to all of the `partials/javascripts/*.html` templates that are loaded and immediately executed.
I chose `main.html` with a block instead of template overrides as it's easier to maintain.

> the handling of the font turned out to be tricky. I would simplify and load the font by default to simplify the example

Good call, adding some `font-extra.css` file to make it load faster (from cache) and just inject the font selector would be the most optimal choice.

> perhaps I would remove the custom CSS bit. Anyone who caters for users that sophisticated probably would not struggle to add this back in?

I will consider it. Though, I do like the idea of a page having the option to just change thet CSS by the user as they see fit. The more options they have to adjust the page for their own preference, the better. People that are tech-savvy in the CSS space, probably use Stylus or some similar browser extension to inject CSS themselves and don't need a setting within the page itself, but beginners might be enticed to try it out.

> perhaps there are other design decisions that would not be necessary for the example?

I will likely skip over the `i18n` part, so there won't ba a need for a hook. As for the simplicity of the JavaScript, I'm still torn between using the current "straight-froward" direct approach of just `+=` concatenating CSS values to a string manually, or if I should write up a more elaborate OOP solution. To be honest I'm not really good at OOP, so I might just add complexity instead of creating a simple interface for the users to add more custom preference toggles.

**Reply by alexvoss** [reply]
> I believe it's needed to inject JavaScript code into the `<head>`, so that it loads the preference before the DOM was completely drawn. 

I see. Does make me wonder if MkDocs should perhaps offer an option to include a small amount if JS in `<head>` instead of just in a separate file. Perhaps there are not enough use cases for this. 

>> remove the custom CSS bit

I would be tempted to reduce the scope and make the example attractive to a wide range of users. User-specified CSS seems to be quite different from simple selectors for fonts and colors.

**Reply by ryran** [reply]
I love what you've got Kamil. Super cool. I'm attempting to adapt it for my site right now.

(I can also absolutely appreciate the idea of simplifying this a bit and getting it into the examples repo.)

**Reply by kamilkrzyskow** [reply]
It kind of fell out of my todo list, as other things stole my attention 😅I thought some more about it and I need some way extra.js scripts can hook into the preferences logic without overriding it, so I'll have to setup at least some OOP with exposed `.add(preferencesHandler)` logic. So it will still take a bit more time.

**Comment by LuchiLucs** [comment]
I'm interested as well, I'm bulding up some sites for local communities and we would like to add a dropdown button to choose between different accessibility fonts. Any update?

**Reply by squidfunk** [reply]
It's currently not planned, as we're working on much bigger things, but this might change in the future. In the meantime, you can use the customization proposed in this discussion.
